### PR TITLE
I6102 ms word cross ref links

### LINKED ISSUE #6102: Treat cross-references in Word as links.
Word can have cross-references - links to other parts of a document (headings, list items, bookmarks, footnotes, etc).  In Word 2016 you create these by pressing alt+s then r then f which opens a dialog box you can setup the exact link from.

The result is a field, code something like:

{ REF _Ref454364297 \h }

which also creates that _Ref454.... bookmark in the appropriate place (a heading in this example)

As this is really a "link", it would be good if NVDA Browse mode could recognise it as a link for navigating.

Note that part of this (and perhaps partly why it isn't already) is that there is currently no keystroke to activate the link - sighted users can hold control and click on the field - as with other links - however enter does not work in this case, it actually doesn't do anything at all and there is currently no other keystroke to activate it.

### PR DESCRIPTION
**Pull request by feerrenrut**
Add support for cross references in word. Fixes #6102 

Now able to report cross references as links when reading through text in word. Can browse by link and when in browse mode a Cross ref link can be activated.

Had to expand the range to the paragraph because single words with in a `ref` dont report as refs. Only the first and the last character.

Refs that are links have the `fields.item(1).code.text` property set to something like `u' REF _Ref457210120 \\h`

For performance reasons the links are cached first, then as we traverse over the words in the range we check if the word is in the vector of known links.

### CODE REVIEW

**Comment by feerrenrut** [comment]
@michaelDCurran would you mind taking a look at this when you have the chance?

File: source/NVDAObjects/window/winword.py [hunk 1]
```diff
@@ -437,8 +445,18 @@ def iterate(self):
...
 	def collectionFromRange(self,rangeObj):
-		return rangeObj.hyperlinks
+		return rangeObj.fields
+
+	def filter(self, item):
+		t = item.type
+		if t == FIELD_TYPE_REF:
+			fieldText = item.code.text.strip().split(' ')
+			# ensure that the text has a \\h in it
+			return any( fieldText[i] == '\\h' for i in range(2, len(fieldText)) )
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/winword.py]: range should be xrange

File: source/NVDAObjects/window/winword.py [hunk 2]
```diff
@@ -499,6 +517,32 @@ def activate(self):
...
+			bookmarkKey = fieldText[1] # we want the _Ref12345 part
+			# get book mark start, we need to look at the whole document to find the bookmark.
+			tempRange.Expand(wdStory)
+			bMark = tempRange.bookmarks(bookmarkKey)
+			self._rangeObj.setRange(bMark.start, bMark.start)
+			self.updateCaret()
+			tiCopy = self.copy()
+			tiCopy.expand(textInfos.UNIT_LINE)
+			speech.speakTextInfo(tiCopy,reason=controlTypes.REASON_FOCUS)
+			braille.handler.handleCaretMove(tiCopy)
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/winword.py]: Should be tempRange (I.e. the collapsed one)

File: source/NVDAObjects/window/winword.py [hunk 3]
```diff
@@ -499,6 +517,32 @@ def activate(self):
...
+			if field.type != FIELD_TYPE_REF:
+				continue
+			fResult = field.result
+			fResult.MoveStart(wdCharacter, -1)
+			if not self._rangeObj.InRange(fResult):
+				continue
+			# text will be something like ' REF _Ref457210120 \\h '
+			fieldText = field.code.text.strip().split(' ')
+			# the \\h field indicates that the field is a link
+			if not any( fieldText[i] == '\\h' for i in range(2, len(fieldText)) ):
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/winword.py]: xrange not range also.

**Comment by michaelDCurran** [comment]
All looks good to me.

**Pull request merged**